File: 0210-course-schedule-ii/0210-course-schedule-ii.cpp
class Solution {
public:
    vector<int> findOrder(int numCourses, vector<vector<int>>& prerequisites) {
        vector<int> answer; 
        queue<int> q; 
        vector<int> inBound(numCourses); 
        vector<vector<int>> adj(numCourses); 

        for(vector<int>& v : prerequisites){
            int to = v[0], from = v[1]; 
            inBound[to]++; 
            adj[from].push_back(to); 
        }

        for(int i = 0; i < numCourses; i++){
            if(inBound[i] == 0){
                q.push(i); 
                answer.push_back(i); 
            }
        }

        int time = 0; 

        while(!q.empty()){
            int size = q.size(); 
            for(int i = 0; i < size; i++){
                int node = q.front(); 
                q.pop(); 
                time++; 
                for(int nextNode : adj[node]){
                    inBound[nextNode]--; 
                    if(inBound[nextNode] == 0){
                        q.push(nextNode); 
                        answer.push_back(nextNode); 
                    }
                }
            }
        }

        if (time != numCourses) return {}; 
        return answer; 
    }
};

File: 0210-course-schedule-ii/0210-course-schedule-ii.java
class Solution {
    public int[] findOrder(int numCourses, int[][] prerequisites) {
        List<Integer> answer = new ArrayList<>(); 
        int[] inOrder = new int[numCourses]; 
        List<List<Integer>> adj = new ArrayList<>(); 
        for(int i = 0; i < numCourses; i++){
            adj.add(new ArrayList<>()); 
        }

        for(int[] next : prerequisites){
            int to = next[0]; 
            int from = next[1]; 

            inOrder[to]++; 
            adj.get(from).add(to); 
        }

        Queue<Integer> q = new LinkedList<>(); 

        for(int i = 0; i < numCourses; i++){
            if(inOrder[i] == 0){
                q.add(i);
            }
        }

        if(q.isEmpty()) return new int[]{}; 


        while(!q.isEmpty()){
            int size = q.size(); 
            for(int i = 0; i < size; i++){
                int node = q.poll();
                answer.add(node); 
                for(int next : adj.get(node)){
                    inOrder[next]--; 
                    if(inOrder[next] == 0){
                        q.add(next); 
                    }
                }
            }
        }

        if(answer.size() != numCourses) return new int[]{}; 
        
        return answer.stream().mapToInt(i -> i).toArray(); 
    }
}